plugins/net_plugin/net_plugin.cpp

```diff
@@ -3143,6 +3143,7 @@ namespace eosio {
    // called from connection strand
    void connection::handle_message( const block_id_type& id, signed_block_ptr ptr ) {
       peer_dlog( this, "received signed_block ${num}, id ${id}", ("num", ptr->block_num())("id", id) );
+      my_impl->producer_plug->received_block();
```

Comment:
I can't remember where we settled on how much validation to perform on a received block before interruption. Should there at least be some sort of check so we don't interrupt on (duplicate) reception of the existing head block?

Reply:
This only interrupts speculative start_block. Any validation would require interacting with the main thread; at least for now. Possibly we could make forkdb thread safe and begin validation before interruption in the future. I see this as a step in the right direction and minimum changes for targeting 3.1 & 3.2.

Reply:
A node may be connected to a handful of nodes & reception of the same block from all those nodes could easily be spread across a wide time window; for each reception it's going to potentially interrupt the speculative start block. On the surface seems like there could be a lot of spurious interruptions?

Reply:
The `received_block()` call is after the de-dup in `net_plugin` `process_next_block_message()`. So only one interruption per unique block_id.

Reply:
Discussed and we want to look at adding block header validation before interrupting speculative `start_block`.

Reply:
Block header validation now performed before interrupting `start_block`.